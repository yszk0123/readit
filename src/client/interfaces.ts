import { Action as ReduxAction } from 'redux';

// Type

export type ID = string;

export type DateOnly = string;

export enum ReviewStatus {
  PLAN_TO_BUY = 'PLAN_TO_BUY',
  UNREAD = 'UNREAD',
  READING = 'READING',
  READ = 'READ',
  STOP = 'STOP',
}

export type ReviewRating = 0 | 1 | 2 | 3 | 4 | 5;

// Entity

export interface User {
  id: ID;
  nickname: string;
  email: string | null;
}

export interface Book {
  id: ID;
  title: string;
  subtitle: string | null;
  author: string;
  thumbnailLink: string;
  description: string;

  category: string;
  pageCount: number;
  publisher: string;
  publishedAt: DateOnly;
  isbn: string | null;
  meta: {
    source: {
      type: string;
      link: string;
      searchText: string;
    };
  };
}

export interface Review {
  id: ID;
  status: ReviewStatus;
  rating: ReviewRating;
}

export interface ReadingLog {
  id: ID;
  book: Book;
  review: Review;
  removed?: boolean;
}

// Action

export interface Action extends ReduxAction {
  payload: any;
}

export enum ActionTypes {
  HELLO = 'HELLO',
  CHANGE_NAME = 'CHANGE_NAME',
  FETCH_SUCCESS = 'FETCH_SUCCESS',
  FETCH_FAILURE = 'FETCH_FAILURE',
  FETCH_READING_LOGS = 'FETCH_READING_LOGS',
  CREATE_READING_LOG = 'CREATE_READING_LOG',
  UPDATE_REVIEW = 'UPDATE_REVIEW',
  SEARCH_BOOK_BY_TITLE = 'SEARCH_BOOK_BY_TITLE',
  REMOVE_READING_LOG = 'REMOVE_READING_LOG',
  REMOVE_READING_LOG_SUCCESS = 'REMOVE_READING_LOG_SUCCESS',
  REMOVE_ENTITY = 'REMOVE_ENTITY',
}

// State

export interface State {
  entities: {
    users: Record<ID, User>;
    books: Record<ID, Book>;
    reviews: Record<ID, Review>;
    readingLogs: Record<ID, ReadingLog>;
  };
  ui: {
    currentUser: ID | null;
    readingLogs: ID[];
    titleInput: string;
  };
}
