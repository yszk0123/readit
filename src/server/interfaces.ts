export interface IModels {
  User: any;
  Book: any;
  Review: any;
}
