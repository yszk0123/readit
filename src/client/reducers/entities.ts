import { State, ActionTypes } from '../interfaces';
import merge from '../utils/merge';
import { Action } from '../actions';

const initialState: State['entities'] = {
  users: {},
  books: {},
  reviews: {},
  readingLogs: {},
};

export default function entities(
  state: State['entities'] = initialState,
  action: Action,
) {
  switch (action.type) {
    case ActionTypes.FETCH_SUCCESS:
      return merge(state, action.payload.entities);
    case ActionTypes.REMOVE_READING_LOG_SUCCESS: {
      const readingLog = state.readingLogs[action.payload.readingLogId];

      return {
        ...state,
        readingLogs: {
          ...state.readingLogs,
          [readingLog.id]: { ...readingLog, removed: true },
        },
      };
    }
    default:
      return state;
  }
}
