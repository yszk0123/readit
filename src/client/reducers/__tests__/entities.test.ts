import { ActionTypes } from '../../interfaces';
import entities from '../entities';

describe('entities', () => {
  it('should merge payload when fetch success', () => {
    const reviewId = '123';
    const review = { id: reviewId };
    const action = {
      type: ActionTypes.FETCH_SUCCESS,
      payload: { entities: { reviews: { [reviewId]: review } } },
    };

    expect(entities(undefined, action)).toMatchSnapshot();
  });
});
