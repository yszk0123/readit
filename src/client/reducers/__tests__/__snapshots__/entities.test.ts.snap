// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entities should merge payload when fetch success 1`] = `
Object {
  "books": Object {},
  "readingLogs": Object {},
  "reviews": Object {
    "123": Object {
      "id": "123",
    },
  },
  "users": Object {},
}
`;
